SYCL: add gelu_erf kernel

qnixsynapse opened the pull request:
*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

qnixsynapse committed:
SYCL: add gelu_erf kernel

ngxson reviewed (approved):
Btw, seems like sycl is missing `GGML_ROPE_TYPE_MROPE`, do you want to give it a try? This type is used by text model, so it will speed up text generation for qwen 2.5

Ref PR where I added it on metal: https://github.com/ggml-org/llama.cpp/pull/13457

qnixsynapse commented:
> Btw, seems like sycl is missing `GGML_ROPE_TYPE_MROPE`, do you want to give it a try? This type is used by text model, so it will speed up text generation for qwen 2.5
> 
> Ref PR where I added it on metal: https://github.com/ggml-org/llama.cpp/pull/13457

Sure. I implemented vision ROPE recently. 

Will open a new PR for that.

AD2605 commented on ggml/src/ggml-sycl/element_wise.cpp:
```diff
@@ -84,6 +84,14 @@ static void gelu_quick(const T *x, T *dst, int k,
+static void gelu_erf(const T * x, T * dst, const int k, const sycl::nd_item<3> &item_ct1) {
+    const T SQRT_2_INV = static_cast<T>(0.70710678118654752440084436210484f);
+    for(auto i = item_ct1.get_global_id(2); i < (const size_t)k; i += item_ct1.get_global_range(2)) {
+        dst[i] = static_cast<T>(0.5f)*x[i]*(static_cast<T>(1.0f) + sycl::erf(x[i]*SQRT_2_INV));
```
```suggestion
       auto x_i = x[i];
        dst[i] = static_cast<T>(0.5f) * x_i * (static_cast<T>(1.0f) + sycl::erf(x_i * SQRT_2_INV));
```

Rbiessy commented on ggml/src/ggml-sycl/element_wise.cpp:
```diff
@@ -1432,6 +1486,12 @@ void ggml_sycl_gelu_quick(ggml_backend_sycl_context & ctx, ggml_tensor * dst) {
 }
 
+void ggml_sycl_gelu_erf(ggml_backend_sycl_context & ctx, ggml_tensor * dst) {
+    GGML_SYCL_DEBUG("call %s: DST Tensor type: %s\n", __func__, ggml_type_name(dst->type));
```
I just merged my dbg print PR, do you mind rebasing and using `scope_op_debug_print`?

qnixsynapse committed:
refactor code

Co-authored-by: Atharva Dubey <atharva.dubey@codeplay.com>

qnixsynapse committed:
Use scope_op_debug_print

AD2605 commented on ggml/src/ggml-sycl/element_wise.cpp:
```diff
@@ -400,6 +408,20 @@ static void gelu_quick_sycl(const T *x, T *dst, const int k,
+                                queue_ptr stream) {
+    const int num_blocks = ceil_div(k, SYCL_GELU_BLOCK_SIZE);
+    stream->parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(1, 1, num_blocks) *
+                              sycl::range<3>(1, 1, SYCL_GELU_BLOCK_SIZE),
+                          sycl::range<3>(1, 1, SYCL_GELU_BLOCK_SIZE)),
```
That sounds okay as well 👍🏻
